ENIGMAsystem/SHELL/Collision_Systems/Polygon/Polygonfuncs.cpp

```diff
@@ -0,0 +1,688 @@
+            const double x1 = x + speed*cos(rad);
+            const double y1 = y - speed*sin(rad);
+            if (collide_inst_inst(object, solid_only, true, x1, y1) != NULL) {
+                break;
+            }
+        }
+        for (d2 = 0; d2 > -radial_max; d2--) // Negative direction.
+        {
+            const double rad = (flipped_direction + d2*angle_radial)*M_PI/180.0;
+            const double x1 = x + speed*cos(rad);
+            const double y1 = y - speed*sin(rad);
+            if (collide_inst_inst(object, solid_only, true, x1, y1) != NULL) {
+                break;
+            }
+        }
+
+        const int d = int(round((d1 + d2)/2.0));
+        const double normal_direction = fmod(flipped_direction + d*angle_radial + 360.0, 360.0);
+
+        // Flip and then mirror the effective direction unit vector along the direction of the normal.
+
+        const double normal_rad = normal_direction * M_PI / 180.0;
+        const double normal_x = cos(normal_rad), normal_y = -sin(normal_rad);
+
+        const double flipped_rad = flipped_direction * M_PI / 180.0;
+        const double flipped_x = cos(flipped_rad), flipped_y = -sin(flipped_rad);
+
+        const double dot_product = normal_x*flipped_x + normal_y*flipped_y;
+        const double flipped_on_normal_x = dot_product*normal_x, flipped_on_normal_y = dot_product*normal_y;
+
+        const double crossed_x = flipped_x - flipped_on_normal_x, crossed_y = flipped_y - flipped_on_normal_y;
+
+        const double mirror_x = flipped_x - 2*crossed_x, mirror_y = flipped_y - 2*crossed_y;
+
+        // Set final direction from flipped, mirrored direction unit vector.
+
+        const double mirror_direction = fmod(atan2(-mirror_y, mirror_x) * 180.0 / M_PI + 360.0, 360.0);
+        inst1->direction = inst1->speed >= 0 ? mirror_direction : fmod(mirror_direction + 180.0, 360.0);
+        return true;
+    }
+    else {
+        const double direction = inst1->speed >= 0 ? inst1->direction : fmod(inst1->direction+180.0, 360.0);
+        const bool free_horizontal = collide_inst_inst(object, solid_only, true, x, y_start) == NULL;
+        const bool free_vertical = collide_inst_inst(object, solid_only, true, x_start, y) == NULL;
+        double new_direction;
+        if (!free_horizontal && free_vertical) {
+            new_direction = direction <= 180.0 ? fmod(180.0 - direction, 360.0) : fmod((360.0 - direction) + 180.0, 360.0);
+        }
+        else if (free_horizontal && !free_vertical) {
+            new_direction = fmod(360.0 - direction, 360.0);
+        }
+        else {
+            new_direction = fmod(direction + 180.0, 360.0);
+        }
+        inst1->direction = inst1->speed >= 0 ? new_direction : fmod(new_direction + 180.0, 360.0);
+
+        return true;
+    }
+}
+
+}
+
+typedef std::pair<int,enigma::object_basic*> inode_pair;
+
+namespace enigma_user
+{
+
+void instance_deactivate_region(int rleft, int rtop, int rwidth, int rheight, bool inside, bool notme) {
```

Comment:
It seems inefficient (maintenance-wise) to duplicate these methods between collision systems... wouldn't surprise me if we were doing this already. I think it would be better to just give collision systems a way to enumerate instances within a region, so we can iterate them inside a main `instance_[de]activate_region` method. Since we're currently not using any broad-phase spatial hashing structures for this, I believe we could just pass an instance_list iterator to the method and have it either wrap the iterator or copy hits into some temp container.